### PR TITLE
rest: clean-up for `mempool` endpoints

### PR DESCRIPTION
**Pull request by brunoerg**
The functions `rest_mempool_info` and `rest_mempool_contents` are similar, the only difference between them is:
`rest_mempool_info` uses `MempoolInfoToJSON` to get the mempool informations and `rest_mempool_contents` uses `MempoolToJSON`, for this reason this PR creates a new function to handle it and reduce duplicated code.

Also,
1. Rename `strURIPart` to `str_uri_part`.
2. Rename `strJSON` to `str_json`.

### CODE REVIEW

**Review by stickies-v — COMMENTED** [review]
Concept ACK 029fad6

Alternatively, you could also only register `/rest/mempool/` and inspect the `str_uri_part`? I think this would allow for slightly more code reduction.

**Comment by brunoerg** [comment]
> Alternatively, you could also only register /rest/mempool/ and inspect the str_uri_part? I think this would allow for slightly more code reduction.

Nice, I had thought about it, but I think `str_uri_part` wouldn't get me the whole URI but only (in this case) the format, like: '.json'. Am I missing something?

**Comment by stickies-v** [comment]
`http_request_cb` [passes](https://github.com/bitcoin/bitcoin/blob/de3c46c93818c6e4000175bcdbf7dd332f54768d/src/httpserver.cpp#L262) everything after the endpoint prefix to `str_uri_part`. If you register the endpoints with prefix `/rest/mempool/<info|content>`, then indeed only `.json` would get passed to `str_uri_part`. If however you just register `/rest/mempool/`, then everything after that i.e. `<info|content>.json` would get passed. (plug: #25754 is trying to clean up this interface)

I don't think the current approach of registering both separately is unreasonable, so this is just a suggestion if you think that implementation makes sense - intuitively it would be my first approach.

**Comment by brunoerg** [comment]
Nice, @stickies-v. Force-pushed addressing your suggestion.

It is clearer having only `/rest/mempool/`, and then check the `str_uri_part`.

**Review by stickies-v — COMMENTED** [review]
Nice! I've left a few more suggestions in comments, but I think that's pretty much it from me. 

<details>
<summary>Couldn't make suggested code, so here's the diff instead:</summary>

```diff
diff --git a/src/rest.cpp b/src/rest.cpp
index dd6fe8d8e..3f2ab98a8 100644
--- a/src/rest.cpp
+++ b/src/rest.cpp
@@ -594,20 +594,23 @@ static bool rest_mempool(const std::any& context, HTTPRequest* req, const std::s
 {
     if (!CheckWarmup(req))
         return false;
-    const CTxMemPool* mempool = GetMemPool(context, req);
-    if (!mempool) return false;
+
     std::string param;
     const RESTResponseFormat rf = ParseDataFormat(param, str_uri_part);
+    if (param != "contents" && param != "info") {
+        return RESTERR(req, HTTP_BAD_REQUEST, "Error: invalid URI format. Expected /rest/mempool/<info|contents>.json");
+    }
+
+    const CTxMemPool* mempool = GetMemPool(context, req);
+    if (!mempool) return false;
 
     switch (rf) {
     case RESTResponseFormat::JSON: {
         std::string str_json;
-        if (str_uri_part == "contents.json") {
+        if (param == "contents") {
             str_json = MempoolToJSON(*mempool, true).write() + "\n";
-        } else if (str_uri_part == "info.json") {
-            str_json = MempoolInfoToJSON(*mempool).write() + "\n";
         } else {
-            return RESTERR(req, HTTP_BAD_REQUEST, "Error: invalid URI format");
+            str_json = MempoolInfoToJSON(*mempool).write() + "\n";
         }
 
         req->WriteHeader("Content-Type", "application/json");
```
</details>

File: src/rest.cpp [hunk 1]
```diff
@@ -590,45 +590,28 @@ static bool rest_chaininfo(const std::any& context, HTTPRequest* req, const std:
...
     case RESTResponseFormat::JSON: {
-        UniValue mempoolObject = MempoolToJSON(*mempool, true);
+        std::string str_json;
+        if (str_uri_part == "contents.json") {
+            str_json = MempoolToJSON(*mempool, true).write() + "\n";
+        } else if (str_uri_part == "info.json") {
+            str_json = MempoolInfoToJSON(*mempool).write() + "\n";
+        } else {
+            return RESTERR(req, HTTP_BAD_REQUEST, "Error: invalid URI format");
+        }
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/rest.cpp:611]: Nice, I agree.

File: src/rest.cpp [hunk 2]
```diff
@@ -590,45 +590,28 @@ static bool rest_chaininfo(const std::any& context, HTTPRequest* req, const std:
...
-    const RESTResponseFormat rf = ParseDataFormat(param, strURIPart);
+    const RESTResponseFormat rf = ParseDataFormat(param, str_uri_part);
 
     switch (rf) {
     case RESTResponseFormat::JSON: {
-        UniValue mempoolObject = MempoolToJSON(*mempool, true);
+        std::string str_json;
+        if (str_uri_part == "contents.json") {
+            str_json = MempoolToJSON(*mempool, true).write() + "\n";
+        } else if (str_uri_part == "info.json") {
```
[the remark this answers is not in the document]
Comment (brunoerg) [reply src/rest.cpp:607]: Done

**Comment by brunoerg** [comment]
Thanks, @stickies-v for your review! Force-pushed addressing the latest suggestions.

File: src/rest.cpp [hunk 3]
```diff
@@ -590,45 +590,31 @@ static bool rest_chaininfo(const std::any& context, HTTPRequest* req, const std:
...
-        req->WriteReply(HTTP_OK, strJSON);
-        return true;
-    }
-    default: {
-        return RESTERR(req, HTTP_NOT_FOUND, "output format not found (available: json)");
-    }
+    std::string param;
+    const RESTResponseFormat rf = ParseDataFormat(param, str_uri_part);
+    if (param != "contents" && param != "info") {
+        return RESTERR(req, HTTP_BAD_REQUEST, "Error: invalid URI format. Expected /rest/mempool/<info|contents>.json");
```
Comment (stickies-v) [inline src/rest.cpp:601]: nit: We already have 3 other instances of this where we don't prefix with "Error:", so maybe best to keep it consistent.
```suggestion
        return RESTERR(req, HTTP_BAD_REQUEST, "Invalid URI format. Expected /rest/mempool/<info|contents>.json");
```

**Review by stickies-v — APPROVED** [review]
ACK da0c612c3d69164da19332167c38bfcd1f9777a8 - nice work on removing all the duplicate code!

Just 1 nit that I didn't see earlier, sorry.

File: src/rest.cpp (hunk 3 above)
Comment (brunoerg) [reply src/rest.cpp:601]: Interesting, there are some places where "Error:" seems to be used. Gonna check it.

Comment (stickies-v) [reply src/rest.cpp:601]: Yeah there is, just not for the "Invalid URI format" ones.

**Review by theStack — APPROVED** [review]
Nice deduplication! 👌 

Code-review ACK da0c612c3d69164da19332167c38bfcd1f9777a8

Happy to also re-ACK if you decide to take [stickies' nit about removing the 'Error' prefix](https://github.com/bitcoin/bitcoin/pull/25760/commits/da0c612c3d69164da19332167c38bfcd1f9777a8#r937008396).

File: src/rest.cpp (hunk 3 above)
Comment (brunoerg) [reply src/rest.cpp:601]: Done!

**Comment by brunoerg** [comment]
Force-pushed removing the `Error:` prefix. Thanks, @stickies-v.

**Review by stickies-v — APPROVED** [review]
re-ACK acbea66589100fe6ef726f4b2a92ec26132ef17b - verified that just the error message was updated since https://github.com/bitcoin/bitcoin/commit/da0c612c3d69164da19332167c38bfcd1f9777a8

**Review by theStack — APPROVED** [review]
re-ACK acbea66589100fe6ef726f4b2a92ec26132ef17b

File: src/rest.cpp [hunk 4]
```diff
@@ -590,45 +590,31 @@ static bool rest_chaininfo(const std::any& context, HTTPRequest* req, const std:
...
 
     switch (rf) {
     case RESTResponseFormat::JSON: {
-        UniValue mempoolObject = MempoolToJSON(*mempool, true);
+        std::string str_json;
+        if (param == "contents") {
+            str_json = MempoolToJSON(*mempool, true).write() + "\n";
+        } else {
+            str_json = MempoolInfoToJSON(*mempool).write() + "\n";
+        }
```
Comment (jonatack) [inline src/rest.cpp:614]: Maybe reduce temporaries and reassignments here.

```diff
     case RESTResponseFormat::JSON: {
-        std::string str_json;
-        if (param == "contents") {
-            str_json = MempoolToJSON(*mempool, true).write() + "\n";
-        } else {
-            str_json = MempoolInfoToJSON(*mempool).write() + "\n";
-        }
-
         req->WriteHeader("Content-Type", "application/json");
-        req->WriteReply(HTTP_OK, str_json);
+        req->WriteReply(HTTP_OK, (param == "contents" ? MempoolToJSON(*mempool, /*verbose=*/true) : MempoolInfoToJSON(*mempool)).write() + "\n");
         return true;
```

File: src/rest.cpp [hunk 5]
```diff
@@ -590,45 +590,31 @@ static bool rest_chaininfo(const std::any& context, HTTPRequest* req, const std:
...
-        std::string strJSON = mempoolInfoObject.write() + "\n";
-        req->WriteHeader("Content-Type", "application/json");
-        req->WriteReply(HTTP_OK, strJSON);
-        return true;
-    }
-    default: {
-        return RESTERR(req, HTTP_NOT_FOUND, "output format not found (available: json)");
-    }
+    std::string param;
+    const RESTResponseFormat rf = ParseDataFormat(param, str_uri_part);
```
Comment (jonatack) [inline src/rest.cpp:599]:
```suggestion
    const RESTResponseFormat& rf = ParseDataFormat(param, str_uri_part);
```

**Review by jonatack — COMMENTED** [review]
Concept ACK

File: src/rest.cpp (hunk 4 above)
Comment (MarcoFalke) [reply src/rest.cpp:614]: It could make sense to avoid `std::string` copies in `WriteReply`, but maybe for a follow-up?

**Pull request merged**